### DISCUSSION TITLE
How could I define the text-overflow method of the DataTable cells?

### DISCUSSION THREAD

**Discussion by Apocryphon-X** [Textualize/textual#3136] [Q&A] [answered]
Hi, as the title says I want to know how to modify the text overflow method of the cells inside a DataTable widget. If I put text larger than the width of the column, the excess characters are simply discarded as if they were not there in the first place.

![image](https://github.com/Textualize/textual/assets/40130428/c86f6272-1c8a-4392-b921-f2fdbffc3386)
> I want to achieve an ellipsis overflow method but I couldn't figure it out how to do it. Like this: ```(Lorem Ipsum Dolor Sit Amet…)```

**Comment by TomJGooding** [comment]
> If I put text larger than the width of the column, the excess characters are simply discarded as if they were not there in the first place.

That's not strictly true, the _value_ of the cell still contains the full contents even if it overflows the visible width (see example below).

Unfortunately looking at how the DataTable is rendered, I'm not sure it is currently possible to display text overflow with ellipsis, or at least not without losing the rest of the content,. But perhaps someone has found a clever workaround!

Alternatively you could set `update_width=True` to resize the column width to accommodate the new cell content.

```python
from textual import on
from textual.app import App, ComposeResult
from textual.widgets import DataTable, Log

LONG_TEXT = "Lorem ipsum dolor sit amet, consectetur adipisci elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua."

COLUMN_WIDTH = 20


class ExampleApp(App):
    CSS = """
    DataTable {
        height: 50%
    }
    Log {
        height: 50%;
        border: double $primary;
    }
    """

    def compose(self) -> ComposeResult:
        yield DataTable()
        yield Log()

    def on_mount(self) -> None:
        log = self.query_one(Log)
        log.border_title = "Highlighted Cell Contents"
        table = self.query_one(DataTable)
        for i in range(3):
            table.add_column(
                f"Column #{i}",
                width=COLUMN_WIDTH,  # Explicitly set the column width
            )
        for _ in range(5):
            table.add_row(*["Example short text" for _ in range(3)])

    @on(DataTable.CellSelected)
    def insert_long_text(self, event: DataTable.CellSelected) -> None:
        event.data_table.update_cell_at(
            event.coordinate,
            LONG_TEXT
            # Alternatively you can resize the column width
            # to accommodate for the new cell content:
            # update_width=True,
        )

    @on(DataTable.CellHighlighted)
    def log_cell(self, event: DataTable.CellHighlighted) -> None:
        log = self.query_one(Log)
        log.write_line(event.data_table.get_cell_at(event.coordinate))


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Reply by willmcgugan** [reply]
I think you might be able to add a Rich Text object rather than plain text, i.e. `Text(LONG_TEXT, overflow="ellipsis")`

**Reply by TomJGooding** [reply]
Thanks Will, I didn't realise Rich Text has an overflow attribute! Unfortunately this doesn't seem to work - see quick test below:

```python
from rich.text import Text
from textual.app import App, ComposeResult
from textual.widgets import DataTable

COLUMN_WIDTH = 25


class ExampleApp(App):
    def compose(self) -> ComposeResult:
        yield DataTable()

    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        table.add_column(
            "Column #1",
            width=COLUMN_WIDTH,
        )
        table.add_row("Example short text")
        table.add_row("Example of long text as string rather than Rich Text")
        table.add_row(
            (
                Text(
                    "Example of long text with Rich Text and ellipsis overflow",
                    overflow="ellipsis",
                )
            )
        )


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Reply by misterek** [reply]
Is there any appetite for a "max_width" option as well?  I use the update_width, but often one really large cell can mess everything up.  

I was thinking `update_width=True`, along with `max_width=50`, or something along those lines.

**Reply by Apocryphon-X** [reply]
Hi everyone! Thanks for the quick answers!

@TomJGooding That idea looks good but sadly I'm the same situation as @misterek, the contents of my cells are usually very long so I can't expand the column in an unlimited way. 

Thinking about the @willmcgugan idea of using a Rich Text, it came to my mind that maybe it's possible to specify a `max_width` with the column size along with the `overflow` method to achieve the desired effect?  I'm not sure how much of this is possible but my idea is something like this: 
```py
Text("Example of long text with Rich Text", overflow="ellipsis", max_width=COLUMN_WIDTH)
```
> Edit: Fixed the double mention to @/misterek.

**Reply by Apocryphon-X** [reply, marked as answer]
In the end I had to do this `long_text[: min(26, len(long_text)) + "…"` as a simple workaround (this may vary depending the use case). With this the cell will indeed have the data cropped out but I made a relation between the cell idx and the original content that was suposed to be there in the first place (this way I'll be able to recover the text without ellipsis).

**Reply by Apocryphon-X** [reply]
I'm marking this as an answer but it wasn't achieved by using the Rich/Textual internal methods though.

**Comment by need47** [comment]
Two years later. I could verify the suggestion from @willmcgugan works:

`Text(LONG_TEXT, overflow="ellipsis", no_wrap=True)`

Note the `no_wrap=True`.
